writer/starlark.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package writer
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// StarlarkWriter is a simple type for writing basic Starlark macros with a consistent form.
+type StarlarkWriter struct {
+	w            *bufio.Writer
+	buf          []string
+	hasBody      bool
+	currentMacro string
+	dirStack     []string
+}
+
+// NewStarlarkWriter creates a new StarlarkWriter writing to the provided output.
+func NewStarlarkWriter(w io.Writer) *StarlarkWriter {
+	return &StarlarkWriter{bufio.NewWriter(w), nil, false, "", nil}
+}
+
+// BeginMacro starts writing a new macro with the given name.
+func (sw *StarlarkWriter) BeginMacro(name string) error {
+	if sw.currentMacro != "" {
+		return errors.New("nested macros are not allowed")
+	}
+	sw.buf = append(sw.buf, fmt.Sprintf("def %s(ctx):\n", name))
+	sw.currentMacro = name
+	sw.hasBody = false
+	return nil
+}
+
+// EndMacro ends writing the current macro; flushing any pending output.
+func (sw *StarlarkWriter) EndMacro() error {
+	if sw.currentMacro == "" {
+		return errors.New("no current macro")
+	}
+	err := sw.writeBuffered()
+	if err != nil {
+		return err
+	}
+	if !sw.hasBody {
+		if err := sw.writeString(sw.indentf("pass\n")); err != nil {
+			return err
+		}
+	}
+	sw.currentMacro = ""
+	return sw.w.Flush()
+}
+
+// PushDirectory writes a Starlark directive indicating a new directory context should be used in the given path.
+func (sw *StarlarkWriter) PushDirectory(path string) error {
+	if sw.currentMacro == "" {
+		return errors.New("no current macro")
+	}
+	sw.dirStack = append(sw.dirStack, path)
+	sw.buf = append(sw.buf, sw.enterString(path))
+	return nil
+}
+
+func (sw *StarlarkWriter) enterString(path string) string {
```

Comment:
Why "enterString"?

Reply:
because it used to be enter/exit directory and pushString sounded too active.  I've changed it to pushDirString